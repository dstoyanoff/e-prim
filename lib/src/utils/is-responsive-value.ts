import { ResponsiveValue } from "../system/types";
import { TBreakpoint } from "../theme";

/**
 * Check if the value is an object that contains either breakpoint
 */
export const isResponsiveValue = <T>(value: ResponsiveValue<T>, breakpoints: TBreakpoint) => {
  if (typeof value !== "object") {
    return false;
  }

  return Object.keys(breakpoints).some(breakpoint => value && breakpoint in value);
};
