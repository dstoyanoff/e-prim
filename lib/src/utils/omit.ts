export const omit = <T extends object>(obj: T, ...keys: (keyof T)[]) => {
  return Object.entries(obj).reduce((acc, [key, value]) => {
    if (!keys.includes(key as keyof T)) {
      acc[key] = value;
    }

    return acc;
  }, {} as Record<string, unknown>) as T;
};
