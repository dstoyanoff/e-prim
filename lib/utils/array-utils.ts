/**
 * Takes either a single element or an array and always returns an array of one or many elements
 * @param items input items of generic type
 */
export const singleOrMulti = <T>(items: T | T[]): T[] => (Array.isArray(items) ? items : [items]);

type ArrayLengthMutationKeys = "splice" | "push" | "pop" | "shift" | "unshift";

export type FixedLengthArray<T, L extends number, TObj = [T, ...Array<T>]> = Pick<
  TObj,
  Exclude<keyof TObj, ArrayLengthMutationKeys>
> & {
  readonly length: L;
  [I: number]: T;
  [Symbol.iterator]: () => IterableIterator<T>;
};
