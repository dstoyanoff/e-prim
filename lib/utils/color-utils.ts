/**
 * Converts a decimal percent to hex
 * @param p percent to convert. Should be an integer between 0 and 100, other values are bounded to that range.
 * Decimal values are rounded to the nearest integer
 */
export const percentToHex = (p: number) => {
  const percent = Math.max(0, Math.min(100, p)); // bound percent from 0 to 100
  const intValue = Math.round((percent / 100) * 255); // map percent to nearest integer (0 - 255)
  const hexValue = intValue.toString(16); // get hexadecimal representation

  return hexValue.padStart(2, "0"); // format with leading 0
};
