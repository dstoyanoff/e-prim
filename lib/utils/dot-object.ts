export type PathsToStringProps<T, TStopType = string> = T extends TStopType
  ? []
  : {
      [K in Extract<keyof T, string | number>]: [
        K,
        ...PathsToStringProps<T[K], TStopType>
      ];
    }[Extract<keyof T, string | number>];

export type Join<
  T extends (string | number)[],
  D extends string = "."
> = T extends []
  ? never
  : T extends [infer F]
  ? F
  : T extends [infer F, ...infer R]
  ? F extends string | number
    ? `${F}${D}${Join<Extract<R, (string | number)[]>, D>}`
    : never
  : string | number;

/**
 * Given a color key (dot-based deep path), resolves the actual color value
 * @param key dot-based path
 * @param object reference to the main object that should be used to find the value
 */
export const getValueFromKey = <
  TValue,
  TKey extends string = string,
  TObject = {}
>(
  key: TKey | undefined,
  object: TObject
): TValue | undefined => {
  if (!key) {
    return undefined;
  }

  const paths = key.split(".");

  return paths.reduce((accumulator, path) => {
    return accumulator ? accumulator[path] : (object as any)[path];
  }, undefined);
};
