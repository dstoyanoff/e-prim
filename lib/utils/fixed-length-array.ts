type ArrayLengthMutationKeys = "splice" | "push" | "pop" | "shift" | "unshift";

export type FixedLengthArray<T, L extends number, TObj = [T, ...Array<T>]> = Pick<
  TObj,
  Exclude<keyof TObj, ArrayLengthMutationKeys>
> & {
  readonly length: L;
  [I: number]: T;
  [Symbol.iterator]: () => IterableIterator<T>;
};
