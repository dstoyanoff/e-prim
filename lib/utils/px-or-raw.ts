export const pxOrRaw = (value: string | number, numberTransform: (value: number) => number = v => v) =>
  typeof value === "number" ? `${numberTransform(value)}px` : value;
