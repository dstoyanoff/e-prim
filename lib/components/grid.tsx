import { ElementType, forwardRef, ReactElement, Ref } from "react";
import { combineResponsiveValues } from "../system/combine-responsive-values";
import { Box, BoxProps } from "./box";
import { omit } from "../utils/omit";
import { createGrid, gridPropKeys, GridSystem } from "../system/grid-system";

const DEFAULT_TAG = "div";

type OwnProps = GridSystem;

export type GridProps<E extends ElementType> = BoxProps<E> & OwnProps;

/**
 * The grid component is a primitive that exposes useful properties for faster prototyping of grid layouts
 * NOTE! This component is in BETA and additional properties will be added in the future
 */
export const Grid: <E extends ElementType = typeof DEFAULT_TAG>(props: GridProps<E>) => ReactElement | null =
  forwardRef(<E extends ElementType = typeof DEFAULT_TAG>(props: GridProps<E>, ref: Ref<GridProps<E>["as"]>) => {
    return (
      <Box
        ref={ref}
        {...omit(props, ...gridPropKeys)}
        css={theme => combineResponsiveValues(...createGrid(props, theme))}
      />
    );
  });
