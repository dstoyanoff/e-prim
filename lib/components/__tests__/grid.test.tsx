/**
 * @jest-environment jsdom
 */

import { render } from "@testing-library/react";
import { wrapper } from "../../utils/mock-theme";
import { Grid } from "../grid";

describe("grid", () => {
  test("should render grid", () => {
    expect(
      render(
        <Grid
          gap={{ xs: 1, md: 3 }}
          columns={{
            xs: [1],
            md: [2, [100, "20%"]],
          }}
          p={3}
          width="100%"
        >
          <div>child1</div>
          <div>child2</div>
        </Grid>,
        {
          wrapper,
        },
      ).container,
    ).toMatchSnapshot();
  });
});
