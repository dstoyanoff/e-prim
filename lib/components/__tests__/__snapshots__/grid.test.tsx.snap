// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`grid should render grid 1`] = `
.emotion-0 {
  width: 100%;
  padding: 12px;
  display: grid;
}

@media (min-width: 0px) {
  .emotion-0 {
    gap: 4px;
    grid-template-columns: repeat(1, 1fr);
  }
}

@media (min-width: 500px) {
  .emotion-0 {
    gap: 12px;
    grid-template-columns: 100px 20%;
  }
}

<div>
  <div
    class="emotion-0"
  >
    <div>
      child1
    </div>
    <div>
      child2
    </div>
  </div>
</div>
`;
